client/schnorr.go

```diff
@@ -9,58 +9,105 @@ import (
 	"math/big"
 )
 
-func (c *Client) Schnorr(protocolType common.ProtocolType, dlog *dlog.ZpDLog, secret *big.Int) error {
-	prover := dlogproofs.NewSchnorrProver(dlog, protocolType)
-	c.handler.schnorrProver = prover
+type SchnorrClient struct {
+	GenericClient
+	prover  *dlogproofs.SchnorrProver
+	secret  *big.Int
+	a       *big.Int
+	variant pb.SchemaVariant
+}
 
-	initMsg := c.getInitialMsg()
+func NewSchnorrClient(endpoint string, protocolType pb.SchemaVariant, dlog *dlog.ZpDLog,
+	s *big.Int) (*SchnorrClient, error) {
+	genericClient, err := NewGenericClient(endpoint)
+	if err != nil {
+		return nil, err
+	}
 
-	if protocolType != common.Sigma {
-		commitment, err := c.open(initMsg) // sends pedersen's h=g^trapdoor
-		if err != nil {
-			return err
-		}
+	return &SchnorrClient{
+		GenericClient: *genericClient,
+		prover:        dlogproofs.NewSchnorrProver(dlog, common.ToProtocolType(protocolType)),
+		secret:        s,
+		a:             dlog.G,
+	}, nil
+}
 
-		c.handler.schnorrProver.PedersenReceiver.SetCommitment(commitment)
-		pedersenDecommitment, err := c.proofRandomData(false, dlog.G, secret)
-		if err != nil {
-			return err
-		}
+// Run calls the appropriate function to execute the Schnorr protocol
+func (c *SchnorrClient) Run() error {
```

Comment:
I like it very much how this function is now all compact and readable, job well done. Clients are now nicely isolated one from each other which make it awesomely readable so 👍